src/decl.cpp

```diff
@@ -141,6 +141,114 @@ static const Type *lApplyTypeQualifiers(int typeQualifiers, const Type *type, So
     return type;
 }
 
+///////////////////////////////////////////////////////////////////////////
+// Attributes
+
+AttrArgument::AttrArgument() : kind(ATTR_ARG_UNKNOWN), intVal(0), stringVal() {}
+AttrArgument::AttrArgument(int64_t i) : kind(ATTR_ARG_UINT32), intVal(i), stringVal() {}
+AttrArgument::AttrArgument(const std::string &s) : kind(ATTR_ARG_STRING), intVal(0), stringVal(s) {}
+
+void AttrArgument::Print() const {
+    switch (kind) {
+    case ATTR_ARG_UINT32:
+        printf("(%" PRId64 ")", intVal);
+        break;
+    case ATTR_ARG_STRING:
+        printf("(\"%s\")", stringVal.c_str());
+        break;
+    case ATTR_ARG_UNKNOWN:
+        printf("(unknown)");
+        break;
+    }
+}
+
+Attribute::Attribute(const std::string &n) : name(n), arg() {}
+Attribute::Attribute(const std::string &n, AttrArgument a) : name(n), arg(a) {}
+Attribute::Attribute(const Attribute &a) : name(a.name), arg(a.arg) {}
+
+bool Attribute::IsKnownAttribute() const {
+    // Known/supported attributes.
+    static std::unordered_map<std::string, int> lKnownParamAttrs = {
+        {"noescape", 1},
+        {"address_space", 1},
+    };
+
+    if (lKnownParamAttrs.find(name) != lKnownParamAttrs.end()) {
+        return true;
+    }
+
+    return false;
+}
+
+void Attribute::Print() const {
+    printf("%s", name.c_str());
+    arg.Print();
+}
+
+AttributeList::AttributeList() {}
+
+AttributeList::AttributeList(const AttributeList &attrList) {
+    auto attrs = attrList.attributes;
+    for (unsigned int i = 0; i < attrs.size(); ++i) {
+        AddAttribute(*attrs[i]);
+    }
+}
+
+AttributeList::~AttributeList() {
+    for (unsigned int i = 0; i < attributes.size(); ++i) {
+        delete attributes[i];
+    }
+}
+
+void AttributeList::AddAttribute(const Attribute &a) {
+    // Create a copy of the given attribute that it owns.
+    Attribute *copy = new Attribute(a);
+    attributes.push_back(copy);
+}
+
+bool AttributeList::HasAttribute(const std::string &name) const {
+    for (unsigned int i = 0; i < attributes.size(); ++i) {
+        if (attributes[i]->name == name)
+            return true;
+    }
+    return false;
+}
+
+Attribute *AttributeList::GetAttribute(const std::string &name) const {
+    for (size_t i = 0; i < attributes.size(); ++i) {
+        if (attributes[i]->name == name)
+            return attributes[i];
+    }
+    return nullptr;
+}
+
+void AttributeList::MergeAttrList(const AttributeList &attrList) {
+    // TODO: consider issuing a warning if the same attribute is specified
+    // several times or with different arguments.
+    for (unsigned int i = 0; i < attrList.attributes.size(); ++i) {
+        Attribute *attr = attrList.attributes[i];
+        if (!HasAttribute(attr->name)) {
+            AddAttribute(*attr);
+        }
+    }
+}
+
+void AttributeList::CheckForUnknownAttributes(SourcePos pos) const {
+    for (size_t i = 0; i < attributes.size(); ++i) {
```

Comment:
Hmm, here you use `size_t`. I suggest being consistent through the patch. My slight preference is range-based for, but `size_t` is also ok.